src/SharpLearning.Backend.Cntk.Test/CntkRawMnistTest.cs

```diff
@@ -0,0 +1,575 @@
+            CntkLayers.Device = device;
+
+            // Define small convolutional neural network.
+            var net = CntkLayers.Input(dataShape);
+
+            net = CntkLayers.Conv2D(net, 3, 3, 32);
+            net = CntkLayers.ActivationFunction(net, Activation.ReLU);
+
+            net = CntkLayers.Pool2D(net, 2, 2);
+
+            net = CntkLayers.Dense(net, 256);
+            net = CntkLayers.ActivationFunction(net, Activation.ReLU);
+
+            net = CntkLayers.Dropout(net, 0.5);
+
+            net = CntkLayers.Dense(net, numberOfClasses);
+            net = CntkLayers.SoftMax(net);
+
+            // Variables for setting features and targets while training the network.
+            var featureVariable = net.Arguments[0];
+            var targetVariable = Variable.InputVariable(new int[] { numberOfClasses }, dataType);
+
+            // Training loss and eval error. Eval error is only for reporting and not used for training.
+            var loss = CNTKLib.CrossEntropyWithSoftmax(net, targetVariable);
+            var evalError = CNTKLib.ClassificationError(net, targetVariable);
+
+            // Setup stochastic gradient descent learner
+            var learningRatePerSample = new TrainingParameterScheduleDouble(0.01, 1);
+            var parameterLearners = new List<Learner>() { Learner.SGDLearner(net.Parameters(), learningRatePerSample) };
+
+            // Setup trainer with the defined network, loss and optimizer.
+            var trainer = Trainer.CreateTrainer(net, loss, evalError, parameterLearners);
+
+            var epochs = 100; // how many epochs to train.
+            var batchSize = 64; // size of each training batch
+            var numberOfBatchesPrEpoch = trainingSampleCount / batchSize; // rough calculation of how many batches pr. epoch
+            var random = new Random(232);
+            var timer = new Stopwatch();
+
+            // Train the model
+            var batchContainer = new Dictionary<Variable, Value>();
+            for (int epoch = 0; epoch < epochs; epoch++)
+            {
+                // Reset accumulations for loss and error metric for each epoch
+                var accumulatedLoss = 0.0;
+                var accumulatedClassificationError = 0.0;
+
+                timer.Restart();
+
+                for (int i = 0; i < numberOfBatchesPrEpoch; i++)
+                {
+                    // Using mnist.GetTrainReader().NextBatchArray(batchSize), which return the data as a single dim array.
+                    // After each epoch, the training data must be shuffled.
+                    // Emulate this by starting the batch at a random index. This is not optimal.
+                    var randomBatchStart = random.Next(0, trainingSampleCount - batchSize);
+                    // the batch images and targets arrays should be reused to reduce allocations.
+                    (var trainingImages, var trainingTargets) = mnist.GetTrainReader(readFromItem: randomBatchStart)
```

Comment:
A hacked way of providing some randomization to order in which the learner/optimizer is presented with new examples. Ideally this should all be handled by the mini-batch source.